crates/matcher/extracted_fzy/src/lib.rs

```diff
@@ -10,6 +10,17 @@ use crate::scoring_utils::*;
 pub type MatchWithPositions = (Score, Vec<usize>);
 
 pub fn match_and_score_with_positions(needle: &str, haystack: &str) -> Option<MatchWithPositions> {
+    let is_case_sensitive = match needle.chars().find(|c| c.is_uppercase()) {
+        Some(_) => true,
+        None => false,
+    };
+    let haystack_opt = if is_case_sensitive {
+        None
+    } else {
+        Some(haystack.to_lowercase())
+    };
+    let haystack = haystack_opt.as_deref().unwrap_or(haystack);
+
```

Comment:
Can be simplified as:

```diff
diff --git a/crates/matcher/extracted_fzy/src/lib.rs b/crates/matcher/extracted_fzy/src/lib.rs
index d69f5da..03292bc 100644
--- a/crates/matcher/extracted_fzy/src/lib.rs
+++ b/crates/matcher/extracted_fzy/src/lib.rs
@@ -10,20 +10,15 @@ use crate::scoring_utils::*;
 pub type MatchWithPositions = (Score, Vec<usize>);

 pub fn match_and_score_with_positions(needle: &str, haystack: &str) -> Option<MatchWithPositions> {
-    let is_case_sensitive = match needle.chars().find(|c| c.is_uppercase()) {
-        Some(_) => true,
-        None => false,
-    };
-    let haystack_opt = if is_case_sensitive {
-        None
+    let haystack = if needle.chars().find(|c| c.is_uppercase()).is_some() {
+        haystack.into()
     } else {
-        Some(haystack.to_lowercase())
+        haystack.to_lowercase()
     };
-    let haystack = haystack_opt.as_deref().unwrap_or(haystack);

-    match matches(needle, haystack) {
+    match matches(needle, &haystack) {
         Some(needle_length) => {
-            let (score, positions) = score_with_positions(needle, needle_length, haystack);
+            let (score, positions) = score_with_positions(needle, needle_length, &haystack);
             Some((score, positions))
         }
         None => None,
```

Reply:
Can you add a Rust test for the smart case?

Reply:
Can you add a Rust test for the smart case?

Reply:
@liuchengxu ok sorry to come back to this and sorry to bother but I have a question that just came to me :sweat_smile:

The way I did it the more complex way (with an `Option`) initially was because from what I read, `String` is allocated on the heap. I was wondering, isn't `haystack.into()` allocating things that would otherwise not be allocating by the way I initially implemented it? Or is it just taking the `&str` without allocating anything on the heap? I still have some trouble following how ownership works and where values are cleared. Thanks.

Reply:
Honstly, I'm also not sure about that 🙈 . After `hayhack.into()`, the ownership was moved, but I'm not sure if the compiler has some optimization tricks to not allocate. The true Rust Expert @ImmemorConsultrixContrarie might have a better answer.

Reply:
This could be simplified a little with a dark magic, but overall `Option` was a good shot.

```rust
let uppercase;
let haystack = if needle.chars().any(|c| c.is_uppercase()) {       
    uppercase = haystack.to_lowercase();
    &uppercase
} else {
    haystack
};
```

Reply:
Great, I had no idea you could create an undefined binding, that's neat :D Thanks to both for the precisions.